### DISCUSSION TITLE
Use of the topics available on the drone

### DISCUSSION THREAD

**Discussion by perezsaura-david** [larics/icuas22_competition#34] [Q&A] [answered]
@lmark1 @fpetric

We have seen in the Github repository documentation that you can control the drone using the following topics: tracker/input_pose, tracker/input_trajectory, and position_hold/trajectory.

Is it also possible to use the rest of the topics, like _mavros_ topics or _carrot_ topics, for example, to control the drone using our own controller?

**Comment by fpetric** [comment, marked as answer]
Hi @perezsaura-david

we are still debating it internally, but we could enable the teams to replace the position controller provided by us. There are two options for this:

1. Remove the position controller, while keeping carrot state machine for reference generation, take offs, lands and so on by commenting out the controller [here](https://github.com/larics/uav_ros_control/blob/17bdc6e4ac021b60e72db9d8c1b4d547d101e245/launch/pid_carrot.launch#L27) and launching your own controller instead. In that case, you need to listen to `nav_msgs/Odometry` on `/red/odometry` and `trajectory_msgs/MultiDofTrajectoryPoint` on `(/red/position_hold/trajectory)`. Your controller should publish `mavros_msgs/AttitudeTarget` to `mavros/setpoint_raw/attitude`.
2. Remove the whole controller and carrot stack by commenting [out this part](https://github.com/larics/icuas22_competition/blob/a9875c685fd93c961c4c1cfbb6d5b169e363fc7c/startup/challenge/session.yml#L31). Again, you need to use odometry topic and so on, but additionally you would need to implement take off service `std_srv/SetBool` `(/red/takeoff)`.

A slight caveat, while we could enable teams to do this for the simulation phase, you will not be able to use your custom controller for the finals, since we do not plan to allow this for a real UAV in any circumstance.

Stay tuned for rulebook updates and announcements here, if we decide to allow teams to implement their controllers for the simulation phase, it will be announced soon.
